File: typescript/tests/cgroup-walker.test.ts
/**
 * Cgroup-scope classifier — Decision #1 of Phase 2 GPU foundation.
 * Mirrors python/tests/test_cgroup_walker.py.
 */

import { describe, it, expect, beforeEach, afterEach } from "vitest";
import {
  mkdtempSync,
  mkdirSync,
  rmSync,
  writeFileSync,
} from "node:fs";
import { join } from "node:path";
import { tmpdir } from "node:os";
import {
  _setCgroupRootForTests,
  _setProcSelfCgroupPathForTests,
  _resetWarningStateForTests,
  classifyScope,
  enumeratePids,
  fallbackLabelFor,
  type CgroupScope,
} from "../src/core/cgroup-walker.js";

function withTmpdir(): { dir: string; setProc: (s: string) => void; cleanup: () => void } {
  const dir = mkdtempSync(join(tmpdir(), "dexcost-cgroup-walker-"));
  const procPath = join(dir, "cgroup");
  _setProcSelfCgroupPathForTests(procPath);
  return {
    dir,
    setProc: (s: string) => writeFileSync(procPath, s),
    cleanup: () => rmSync(dir, { recursive: true, force: true }),
  };
}

describe("classifyScope — Decision #1 table", () => {
  let tmp: ReturnType<typeof withTmpdir>;
  beforeEach(() => {
    _resetWarningStateForTests();
    tmp = withTmpdir();
  });
  afterEach(() => {
    tmp.cleanup();
    _setProcSelfCgroupPathForTests(null);
    _setCgroupRootForTests(null);
  });

  const table: [string, string, string][] = [
    ["docker", "0::/docker/abc123\n", "container"],
    ["kubepods.slice", "0::/kubepods.slice/kubepods-burstable.slice/foo.scope\n", "container"],
    ["kubepods (legacy)", "0::/kubepods/burstable/podabc/abc\n", "container"],
    ["system.slice docker-", "0::/system.slice/docker-abc.scope\n", "container"],
    ["system.slice containerd-", "0::/system.slice/containerd-abc.scope\n", "container"],
    ["system.slice crio-", "0::/system.slice/crio-abc.scope\n", "container"],
    ["containerd", "0::/containerd/abc\n", "container"],
    ["crio", "0::/crio/abc\n", "container"],
    ["user.slice", "0::/user.slice/user-1000.slice/session-2.scope\n", "bare_metal_user_slice"],
    [
      "user.slice deep",
      "0::/user.slice/user-1000.slice/user@1000.service/app.slice/unit.service\n",
      "bare_metal_user_slice",
    ],
    ["root", "0::/\n", "root_cgroup"],
    ["unknown prefix", "0::/some/unknown/path\n", "unknown"],
  ];

  it.each(table)("%s → %s", (_name, content, expected) => {
    tmp.setProc(content);
    expect(classifyScope().kind).toBe(expected);
  });

  it("multi-line file → cgroup_v1", () => {
    tmp.setProc(
      "12:devices:/docker/abc\n11:cpuset:/docker/abc\n10:memory:/docker/abc\n",
    );
    expect(classifyScope().kind).toBe("cgroup_v1");
  });

  it("missing /proc/self/cgroup → unknown", () => {
    _setProcSelfCgroupPathForTests("/nonexistent/path/cgroup");
    const scope = classifyScope();
    expect(scope.kind).toBe("unknown");
    expect(scope.path).toBe(null);
  });

  it("container scope carries the resolved path", () => {
    tmp.setProc("0::/kubepods.slice/kubepods-burstable.slice/foo.scope\n");
    const scope = classifyScope();
    expect(scope.path).toBe(
      "/kubepods.slice/kubepods-burstable.slice/foo.scope",
    );
  });
});

describe("enumeratePids — container walks; non-container returns self-PID only", () => {
  let tmp: ReturnType<typeof withTmpdir>;
  beforeEach(() => {
    _resetWarningStateForTests();
    tmp = withTmpdir();
  });
  afterEach(() => {
    tmp.cleanup();
    _setProcSelfCgroupPathForTests(null);
    _setCgroupRootForTests(null);
  });

  it("container scope walks /sys/fs/cgroup/<path>/cgroup.procs", () => {
    tmp.setProc("0::/docker/abc123\n");
    const root = join(tmp.dir, "fake_cgroup_root");
    mkdirSync(join(root, "docker", "abc123"), { recursive: true });
    writeFileSync(
      join(root, "docker", "abc123", "cgroup.procs"),
      "1234\n5678\n9012\n",
    );
    _setCgroupRootForTests(root);

    const scope = classifyScope();
    const pids = enumeratePids(scope);
    expect(pids).toEqual([1234, 5678, 9012]);
  });

  it("bare_metal_user_slice → [self_pid] (silent-overcount guard)", () => {
    tmp.setProc("0::/user.slice/user-1000.slice/session-2.scope\n");
    const scope = classifyScope();
    expect(enumeratePids(scope)).toEqual([process.pid]);
  });

  it("root_cgroup → [self_pid] (ambiguous)", () => {
    tmp.setProc("0::/\n");
    expect(enumeratePids(classifyScope())).toEqual([process.pid]);
  });

  it("unknown scope → [self_pid]", () => {
    tmp.setProc("0::/some/unknown/path\n");
    expect(enumeratePids(classifyScope())).toEqual([process.pid]);
  });

  it("cgroup_v1 → [self_pid] (v1.1 deferred)", () => {
    tmp.setProc("12:devices:/docker/abc\n11:cpuset:/docker/abc\n");
    expect(enumeratePids(classifyScope())).toEqual([process.pid]);
  });

  it("container walk denied → null + warn-once", () => {
    tmp.setProc("0::/docker/abc123\n");
    _setCgroupRootForTests(join(tmp.dir, "nonexistent"));
    const scope = classifyScope();
    expect(enumeratePids(scope)).toBeNull();
  });
});

describe("fallbackLabelFor — Decision #1 pricing-source suffix", () => {
  it("container → null (no fallback)", () => {
    const s: CgroupScope = { kind: "container", path: "/docker/abc" };
    expect(fallbackLabelFor(s)).toBeNull();
  });
  it("bare_metal_user_slice → no_container_scope", () => {
    expect(fallbackLabelFor({ kind: "bare_metal_user_slice", path: null })).toBe(
      "no_container_scope",
    );
  });
  it("root_cgroup → no_container_scope", () => {
    expect(fallbackLabelFor({ kind: "root_cgroup", path: null })).toBe(
      "no_container_scope",
    );
  });
  it("unknown → self_pid_only", () => {
    expect(fallbackLabelFor({ kind: "unknown", path: null })).toBe(
      "self_pid_only",
    );
  });
  it("cgroup_v1 → self_pid_only", () => {
    expect(fallbackLabelFor({ kind: "cgroup_v1", path: null })).toBe(
      "self_pid_only",
    );
  });
});

File: go/core/cgroup_walker_test.go
// Task 2 — cgroup-scope classifier tests. Mirrors python commit caebcf7.
//
// The classifier is the Decision #1 verification-gate boundary: it inspects
// /proc/self/cgroup, identifies the scope, and tells GpuAccountant whether
// to walk the cgroup's PIDs or degrade to self-PID-only.

package core

import (
	"os"
	"path/filepath"
	"strconv"
	"testing"
)

// writeCgroupFile creates a fake /proc/self/cgroup-style file.
func writeCgroupFile(t *testing.T, content string) string {
	t.Helper()
	dir := t.TempDir()
	path := filepath.Join(dir, "cgroup")
	if err := os.WriteFile(path, []byte(content), 0644); err != nil {
		t.Fatalf("write fake cgroup file: %v", err)
	}
	return path
}

func TestClassifyScopeContainerKubepodsSlice(t *testing.T) {
	procPath := writeCgroupFile(t,
		"0::/kubepods.slice/kubepods-burstable.slice/kubepods-burstable-pod1234.slice/cri-containerd-abcd.scope\n")
	scope := classifyCgroupScopeFromFile(procPath)
	if scope.Kind != CgroupKindContainer {
		t.Fatalf("kubepods.slice should classify as container; got %q", scope.Kind)
	}
	if scope.Path == "" {
		t.Fatalf("container scope should have a non-empty Path")
	}
}

func TestClassifyScopeContainerPrefixes(t *testing.T) {
	// All seven container-prefix variants from Decision #1's table.
	cases := []struct{ input string }{
		{"0::/kubepods.slice/kubepods-pod1.slice\n"},
		{"0::/kubepods/burstable/pod1\n"},
		{"0::/docker/abc123\n"},
		{"0::/system.slice/docker-abc.scope\n"},
		{"0::/containerd/abc\n"},
		{"0::/system.slice/containerd-abc.scope\n"},
		{"0::/crio/abc\n"},
		{"0::/system.slice/crio-abc.scope\n"},
	}
	for _, c := range cases {
		path := writeCgroupFile(t, c.input)
		scope := classifyCgroupScopeFromFile(path)
		if scope.Kind != CgroupKindContainer {
			t.Errorf("%q should classify as container; got %q", c.input, scope.Kind)
		}
	}
}

func TestClassifyScopeBareMetalUserSlice(t *testing.T) {
	procPath := writeCgroupFile(t,
		"0::/user.slice/user-1000.slice/session-1.scope\n")
	scope := classifyCgroupScopeFromFile(procPath)
	if scope.Kind != CgroupKindBareMetalUserSlice {
		t.Fatalf("user.slice should classify as bare_metal_user_slice; got %q", scope.Kind)
	}
	if scope.Path != "" {
		t.Fatalf("bare_metal scope should have empty Path; got %q", scope.Path)
	}
}

func TestClassifyScopeRootCgroup(t *testing.T) {
	procPath := writeCgroupFile(t, "0::/\n")
	scope := classifyCgroupScopeFromFile(procPath)
	if scope.Kind != CgroupKindRootCgroup {
		t.Fatalf("root cgroup should classify as root_cgroup; got %q", scope.Kind)
	}
}

func TestClassifyScopeCgroupV1MultiLine(t *testing.T) {
	procPath := writeCgroupFile(t,
		"12:devices:/docker/abc\n"+
			"11:cpu,cpuacct:/docker/abc\n"+
			"10:memory:/docker/abc\n")
	scope := classifyCgroupScopeFromFile(procPath)
	if scope.Kind != CgroupKindCgroupV1 {
		t.Fatalf("multi-line should classify as cgroup_v1; got %q", scope.Kind)
	}
}

func TestClassifyScopeUnknown(t *testing.T) {
	procPath := writeCgroupFile(t, "0::/something/weird\n")
	scope := classifyCgroupScopeFromFile(procPath)
	if scope.Kind != CgroupKindUnknown {
		t.Fatalf("unmatched path should classify as unknown; got %q", scope.Kind)
	}
}

func TestClassifyScopeMissingFile(t *testing.T) {
	scope := classifyCgroupScopeFromFile("/no/such/file/anywhere")
	if scope.Kind != CgroupKindUnknown {
		t.Fatalf("missing file should fall back to unknown; got %q", scope.Kind)
	}
}

func TestEnumeratePIDsBareMetalReturnsSelfPID(t *testing.T) {
	scope := CgroupScope{Kind: CgroupKindBareMetalUserSlice}
	pids := EnumerateCgroupPIDs(scope, "")
	if len(pids) != 1 || pids[0] != os.Getpid() {
		t.Fatalf("bare-metal should degrade to self-PID-only; got %v", pids)
	}
}

func TestEnumeratePIDsRootCgroupReturnsSelfPID(t *testing.T) {
	scope := CgroupScope{Kind: CgroupKindRootCgroup}
	pids := EnumerateCgroupPIDs(scope, "")
	if len(pids) != 1 || pids[0] != os.Getpid() {
		t.Fatalf("root cgroup should degrade to self-PID-only; got %v", pids)
	}
}

func TestEnumeratePIDsCgroupV1ReturnsSelfPID(t *testing.T) {
	scope := CgroupScope{Kind: CgroupKindCgroupV1}
	pids := EnumerateCgroupPIDs(scope, "")
	if len(pids) != 1 || pids[0] != os.Getpid() {
		t.Fatalf("cgroup_v1 should degrade to self-PID-only; got %v", pids)
	}
}

func TestEnumeratePIDsContainerWalksProcsFile(t *testing.T) {
	// Fake cgroup root with a cgroup.procs file inside.
	root := t.TempDir()
	containerPath := "/kubepods.slice/kubepods-pod1.slice"
	dir := filepath.Join(root, containerPath)
	if err := os.MkdirAll(dir, 0755); err != nil {
		t.Fatalf("mkdir: %v", err)
	}
	procsFile := filepath.Join(dir, "cgroup.procs")
	content := "1234\n5678\n9012\n"
	if err := os.WriteFile(procsFile, []byte(content), 0644); err != nil {
		t.Fatalf("write procs: %v", err)
	}

	scope := CgroupScope{Kind: CgroupKindContainer, Path: containerPath}
	pids := EnumerateCgroupPIDs(scope, root)
	want := []int{1234, 5678, 9012}
	if len(pids) != len(want) {
		t.Fatalf("PIDs len: got %v; want %v", pids, want)
	}
	for i, w := range want {
		if pids[i] != w {
			t.Fatalf("PID[%d] = %d; want %d", i, pids[i], w)
		}
	}
}

func TestEnumeratePIDsContainerReadFailureReturnsNil(t *testing.T) {
	// container scope but no cgroup.procs at path → nil (caller degrades).
	scope := CgroupScope{Kind: CgroupKindContainer, Path: "/kubepods.slice/nope"}
	pids := EnumerateCgroupPIDs(scope, t.TempDir())
	if pids != nil {
		t.Fatalf("expected nil when cgroup.procs unreadable; got %v", pids)
	}
}

// ─── Decision #1 fallback labels ───────────────────────────────────────

func TestFallbackLabelContainerReturnsEmpty(t *testing.T) {
	if l := FallbackLabelForScope(CgroupScope{Kind: CgroupKindContainer}); l != "" {
		t.Fatalf("container should have no fallback label; got %q", l)
	}
}

func TestFallbackLabelBareMetalReturnsNoContainerScope(t *testing.T) {
	if l := FallbackLabelForScope(CgroupScope{Kind: CgroupKindBareMetalUserSlice}); l != "no_container_scope" {
		t.Fatalf("bare_metal label = %q; want no_container_scope", l)
	}
	if l := FallbackLabelForScope(CgroupScope{Kind: CgroupKindRootCgroup}); l != "no_container_scope" {
		t.Fatalf("root_cgroup label = %q; want no_container_scope", l)
	}
}

func TestFallbackLabelCgroupV1Unknown(t *testing.T) {
	if l := FallbackLabelForScope(CgroupScope{Kind: CgroupKindCgroupV1}); l != "self_pid_only" {
		t.Fatalf("cgroup_v1 label = %q; want self_pid_only", l)
	}
	if l := FallbackLabelForScope(CgroupScope{Kind: CgroupKindUnknown}); l != "self_pid_only" {
		t.Fatalf("unknown label = %q; want self_pid_only", l)
	}
}

// Sanity: prefix table is exhaustive (no missing common Kubernetes shape).
func TestCgroupPrefixTableIsComplete(t *testing.T) {
	expectedPrefixes := []string{
		"/kubepods.slice/", "/kubepods/", "/docker/",
		"/system.slice/docker-", "/containerd/",
		"/system.slice/containerd-", "/crio/",
		"/system.slice/crio-",
	}
	for _, p := range expectedPrefixes {
		found := false
		for _, actual := range containerPrefixes {
			if actual == p {
				found = true
				break
			}
		}
		if !found {
			t.Errorf("prefix table missing %q (Decision #1)", p)
		}
	}
}

// Don't trip ourselves by silently swallowing non-int lines in cgroup.procs.
func TestEnumeratePIDsContainerIgnoresMalformedLines(t *testing.T) {
	root := t.TempDir()
	containerPath := "/docker/abc"
	dir := filepath.Join(root, containerPath)
	_ = os.MkdirAll(dir, 0755)
	_ = os.WriteFile(filepath.Join(dir, "cgroup.procs"),
		[]byte("1234\nnot-a-pid\n5678\n\n"), 0644)

	scope := CgroupScope{Kind: CgroupKindContainer, Path: containerPath}
	pids := EnumerateCgroupPIDs(scope, root)
	if len(pids) != 2 {
		t.Fatalf("expected 2 valid PIDs; got %v", pids)
	}
	if strconv.Itoa(pids[0]) != "1234" || strconv.Itoa(pids[1]) != "5678" {
		t.Fatalf("PIDs unexpected: %v", pids)
	}
}
